### PR TITLE
Driver interface refactor

### LINKED ISSUE #743: Define Driver Standard Interface
This will define the standard driver interface to be used in all drivers.   Most of it is already defined but this will finish it so we can merge the new drivers.

Refs #554

### PR DESCRIPTION
**Pull request by ehazlett**
This adds a few helpers for SSH so we can remove the ssh operations from provisioning.  It also removes the Docker specific operations like starting / stopping the daemon and `GetSSHCommand` which will be refactored in #744.

Closes #743

### CODE REVIEW

File: drivers/drivers.go [hunk 1]
```diff
@@ -14,62 +13,57 @@ import (
...
 
-	// SetConfigFromFlags configures the driver with the object that was returned
-	// by RegisterCreateFlags
-	SetConfigFromFlags(flags DriverOptions) error
+	// GetIP returns an IP or hostname that this host is available at
+	// e.g. 1.2.3.4 or docker-host-d60b70a14d3a.cloudapp.net
+	GetIP() (string, error)
+
+	// GetSSHHostname returns hostname for use with ssh
+	GetSSHHostname() (string, error)
```
Comment (sthulb) [inline drivers/drivers.go]: I'd convert this to `GetSSHAddr(ess)` it's rare we actually use hostnames, and addr is generic.

Comment (ehazlett) [inline drivers/drivers.go]: Good point.  Thanks!

File: drivers/drivers.go [hunk 2]
```diff
@@ -3,73 +3,78 @@ package drivers
...
 	DriverName() string
 
-	// SetConfigFromFlags configures the driver with the object that was returned
-	// by RegisterCreateFlags
-	SetConfigFromFlags(flags DriverOptions) error
+	// GetIP returns an IP or hostname that this host is available at
+	// e.g. 1.2.3.4 or docker-host-d60b70a14d3a.cloudapp.net
+	GetIP() (string, error)
+
+	// GetSSHHostname returns hostname for use with ssh
```
Comment (nathanleclaire) [inline drivers/drivers.go]: The comment says "GetSSHHostname", but it is named `GetSSHAddress` here

**Comment by nathanleclaire** [comment]
It would probably be faster to split my existing PR () out into a few different pieces, one of which would be this piece, the new SSH functions.  I'd happily split it into a few different PRs if you want.

**Comment by ehazlett** [comment]
@nathanleclaire yeah that sounds great -- especially the ssh and waitFunc stuff.  I'm mainly focusing on the driver specific implementation to match this updated interface.  thanks!

File: drivers/drivers.go (hunk 2 above)
Comment (ehazlett) [inline drivers/drivers.go]: Oops - good catch.  Thanks!

**Comment by nathanleclaire** [comment]
OK - I will put in PRs for those (maybe pulling out the hostname provisioning as well)

**Comment by ehazlett** [comment]
@nathanleclaire i have the hostname stuff coming -- can you comment on it first?  i want to see what you think.

**Comment by ehazlett** [comment]
@nathanleclaire fyi using `SaveConfig` after provisioning is making working with ec2 much much better :)  :+1:

**Comment by nathanleclaire** [comment]
Nice.

Mostly looks pretty good, a few things to consider:

- I'd favor `GetSSHHostname` over `GetSSHAddress` - part of the reason this method needs to be implemented is to get a reliable hostname to run SSH commands against, one of which is `localhost` (in b2d/vbox's case)
- `HypervisorType` strikes me as a little misleading.  Types of hypervisors are ESXI, Xen, KVM, etc.  But at the same time I'm struggling to come up with something better, so it might be okay.

**Comment by ehazlett** [comment]
@nathanleclaire hmm ok -- we do return addresses for most of those.  is there a reason we return "localhost" and not "127.0.0.1" or something?

I agree with HypervisorType -- I was in the same boat -- not sure what to call it.  I had a bool just returning IsLocal but that seemed a bit restrictive.

**Comment by ehazlett** [comment]
Maybe `ProviderType` with `Local`, `Remote`?

**Comment by nathanleclaire** [comment]
> hmm ok -- we do return addresses for most of those. is there a reason we return "localhost" and not "127.0.0.1" or something?

Hmmmm either way will probably work but doing it by hostname was @SvenDowideit's suggestion, so perhaps he can comment on it.

`ProviderType` sounds okay - although thinking about it, would you be opposed to adding a basic `Provisioner` interface, doing detection of the underlying OS (only Ubuntu and b2d implementations supported for now) and then using a variation of that interface `Boot2DockerProvisioner` or `UbuntuProvisioner`?  This would replace the `switch` statements with just one method call, and IMO get us closer to our eventual goal of being able to pull provisioning out of the drivers completely instead of asking them to self-report how they should be provisioned.

**Comment by ehazlett** [comment]
@nathanleclaire yeah the idea was to get a Provisioner (#699).  I'm not opposed to that.  You want to do a PR with that?

**Comment by sthulb** [comment]
I can finish my OS detection PR if it'll make #699 easier.

**Comment by ehazlett** [comment]
@sthulb yeah that sounds great. thanks!

**Comment by ehazlett** [comment]
@nathanleclaire for now I will change `HypervisorType` to `ProviderType` until we address with the provisioner.  I think we will still probably need to have the Driver inform what type of provider they are, no?

**Comment by nathanleclaire** [comment]
@ehazlett It probably can't hurt to have the methods available so that SGTM.  Maybe I can coordinate with @sthulb to shape up the `Provisioner` stuff we're discussing too.

**Comment by ehazlett** [comment]
@nathanleclaire perfect.  sounds great.  thanks!

**Comment by sthulb** [comment]
@nathanleclaire catchup tomorrow?

**Comment by ehazlett** [comment]
PTAL @sthulb @nathanleclaire

**Comment by nathanleclaire** [comment]
Looking pretty good... I'll pull it down locally and run through some testing with it!

File: drivers/azure/azure.go [hunk 3]
```diff
@@ -51,7 +52,7 @@ func init() {
 	})
 }
 
-// GetCreateFlags registers the flags this driver adds to
+// GetCreateFlags registers the flags this d adds to
```
Comment (ahmetalpbalkan) [inline drivers/azure/azure.go]: this doesn't seem like fixed?

Comment (ehazlett) [inline drivers/azure/azure.go]: Yeah I need to merge in master.  Doing that now.  Thanks!

**Comment by nathanleclaire** [comment]
I thought I commented but so far in my testing I've tried DigitalOcean, VirtualBox, and Amazon EC2 and they all work great.  This is looking really good from my end.

@ehazlett Is your plan for dealing with breakage of older machines that lack the expected SSH fields still the same (default to 22 etc.)?  And do you plan to implement it in this PR, or in a follow-up?

**Comment by sthulb** [comment]
Arse. My comment from yesterday didn't go through apparently. 

I've tried VMware + EC2, looking good so far :+1:

**Comment by ehazlett** [comment]
@sthulb thx!

**Comment by ehazlett** [comment]
@nathanleclaire i want to address the ssh port/user issue in this one too.  working on it today.  thx!

**Comment by ehazlett** [comment]
@nathanleclaire ok so i have the `SSHUser` and `SSHPort` issues fixed.  They are not ideal as I updated each driver, however, I think it's the safest.  Each driver can specify their own usernames / ports and I think it's up to them to make sure they are correct.  I started with a central function in host that would check if missing but it seemed dirty having the logic there.  I think this puts the responsibility where it should be and resolves the issue.

What this also brought back up in my head was the need to version the configs but that can come in a separate PR :)

**Comment by nathanleclaire** [comment]
Yeah, let's not get too ahead of ourselves.  I really want to get this PR in and start banging on it in combination with some of the other stuff we have in the works.

LGTM

**Comment by sthulb** [comment]
Small PRs = Good :)

LGTM

**Comment by ehazlett** [comment]
Cool thx!

**Pull request merged**